### PR TITLE
Release v1.0.0

### PR DESCRIPTION
**Pull request by hollodotme**
- Adds `Abstract` prefix to type classes
- Adds `4` as UUID version indicator to `AbstractUuidType`
- Adds `abstract protected guardValueIsValid(<type> $value)` to all basic types
- Adds tests for new guard methods
- Updates changelog and documentation

### CODE REVIEW

**Comment by mfuehrer82** [comment]
Make constructors of abstract classes final to prevent overwrite in concrete types.

**Comment by hollodotme** [comment]
@mfuehrer82 I would argue that this limits the usage of types. For example if you want to have a type which consists of e.g. two string values, but can be represented as one string internally, like `FullName`, it would not be possible to create a constructor for it.

Example:

```php
<?php

final class FullName extends AbstractStringType
{
    public function __construct(string $firstname, string $lastname)
    {
        parent::__construct($firstname . ' ' . $lastname);
    }

    protected function guardValueIsValid( string $value )
    {
        if (false === strpos($value, ' '))
        {
            throw new InvalidArgumentException('A full name must have at least one whitespace.');
        }
    }

    public static function fromString( string $value ) : FullName
    {
        $names = explode(' ', $value);

        return new self(implode(' ', array_slice($names, 0, -1), end($names));
    }
}
```

File: README.md [hunk 1]
```diff
@@ -19,32 +21,41 @@ composer require fortuneglobe/types
...
 namespace YourVendor\YourProject;
 
-use Fortuneglobe\Types\StringType;
+use Fortuneglobe\Types\AbstractStringType;
+use Fortuneglobe\Types\Exceptions\InvalidArgumentException;
 
-final class UserId extends StringType {}
+final class UserId extends AbstractStringType 
+{
+    protected function guardValueIsValid( string $value )
```
Comment (codekandis) [inline README.md]: The return type is omitted.

```php
protected function guardValueIsValid( string $value ): void
```

Regarding to this the guard methods in all abstract types need to be fixed.

File: src/AbstractFloatType.php [hunk 2]
```diff
@@ -6,19 +6,23 @@
...
 	private $value;
 
 	public function __construct( float $value )
 	{
+		$this->guardValueIsValid( $value );
+
 		$this->value = $value;
 	}
 
+	abstract protected function guardValueIsValid( float $value );
```
Comment (codekandis) [inline src/AbstractFloatType.php]: The return type is omitted.

```php
abstract protected function guardValueIsValid( float $value ): void
```

Regarding to this the guard methods in all abstract types need to be fixed.

File: README.md [hunk 3]
```diff
@@ -19,32 +21,41 @@ composer require fortuneglobe/types
...
 ### Available base types
 
-* StringType
-* FloatType
-* IntType
-* UuidType (extends StringType)
+* AbstractStringType
+* AbstractFloatType
+* AbstractIntType
+* AbstractUuidType (extends AbstractStringType)
```
Comment (codekandis) [inline README.md]: It's `AbstractUuid4Type`.

File: tests/Unit/AbstractUuid4TypeTest.php [hunk 4]
```diff
@@ -2,26 +2,26 @@
...
-use Fortuneglobe\Types\UuidType;
+use Fortuneglobe\Types\Interfaces\RepresentsUuid4Value;
+use Fortuneglobe\Types\Tests\Unit\Fixtures\TestUuid4;
 use PHPUnit\Framework\TestCase;
 
 /**
  * Class UuidTypeTest
  * @package Fortuneglobe\Types\Tests\Unit
  */
-final class UuidTypeTest extends TestCase
```
Comment (codekandis) [inline tests/Unit/AbstractUuid4TypeTest.php]: It's `AbstractUuid4TypeTest`.

**Review by codekandis — CHANGES_REQUESTED** [review]
There are several return types inconsistant or completely missing in the test cases.

File: README.md (hunk 1 above)
Comment (hollodotme) [reply README.md]: Fixed in 70fe97374eeed55e6dce6d5d1872fd2f9ec9974c

File: src/AbstractFloatType.php (hunk 2 above)
Comment (hollodotme) [reply src/AbstractFloatType.php]: Fixed in 70fe97374eeed55e6dce6d5d1872fd2f9ec9974c

File: tests/Unit/AbstractUuid4TypeTest.php (hunk 4 above)
Comment (hollodotme) [reply tests/Unit/AbstractUuid4TypeTest.php]: Fixed in 70fe97374eeed55e6dce6d5d1872fd2f9ec9974c

File: README.md (hunk 3 above)
Comment (hollodotme) [reply README.md]: Fixed in 1fcd1e80b61d3f6650e3ed11cc2b2d3b3deb4f0b

**Review by codekandis — APPROVED** [review]

**Pull request closed without merging**